broadinstitute/wfl
# GH-655: WFL-Cromwell communication

### Purpose


- https://broadinstitute.atlassian.net/browse/GH-655

### Changes


- Implement the function to generate `GoogleCredentials` objects and `AccessToken` from service account files, this is also taking the deployment switch from GAE to GKE into consideration. The function depends on a few environment variables defined in the `server.clj`.
- Update ~a few (but not all)~ ALL  of handlers to use the new way of authentication.
- Fix a security alert for the UI.
- Clean up some testing comments in the Helm chart files. 
- **Removed all about OAUTH from the API server. ** (DO NOT deploy this until the login page for the UI is implemented and merged.)
- Update the Helm chart for WFL to include a secret object that stores service account.
- **PR https://github.com/broadinstitute/pipeline-config/pull/25 needs to be merged before this one and we shall update the commit hash in `boot.clj` before merging this one.**

## Open Questions

- [x] This PR does not remove the `authorize` handler wrapper, which only validates the user provided token, decode into JWTs and store them into the ring session's request body. According to @tbl3rd and @hjfbynara we will likely to get the user info(email) from the OIDC proxy in the future, so should we remove this wrapper now or later once we move to GKE? Also, 
is `get the user info(email) from the OIDC proxy` true?
- [x] There are 2 ways of passing the auth tokens:
    - [x] top down: generate the token and passing them from the handlers all the way down to the underlying functions
    - [x] bottom up: generate the token from the underlying functions, basically replace all of the places that are currently using `once/get-local-auth-header`.
  This PR updates a few handlers using the first approach, but it's very likely we need to do a lot with the second approach. Question is it will **break the WFL CLI**, shall we stick with the 1st approach, even it's much more complicated? (Requires refactor a lot of functions to support polymorphism)
- [x] Currently a lot of GCloud service to service communication are done using `once/get-local-auth-header` as well, leveraging the fact that GAE handles some authentication for us. Moving to GKE will change this and we will need to do a lot by ourselves. How could we do it? (The 2nd approach above seems to be more appealing taking this into consideration)

**[Updated on 05/01]: All open questions are resolved now thanks to @tbl3rd and @samanehsan** 


### Review Instructions / QA


- There is a sister PR ready for review as well: https://github.com/broadinstitute/pipeline-config/pull/25
~- Note: this PR won't be **Finished** until we move to GKE completely, so please don't expect this PR to be fully QAed by then!~ We have moved to GKE, which is done in another PR: #23 

1. Make sure you are on the list of `vault read secret/dsde/gotc/dev/wfl/users`, if not, add yourself with `curl --silent -H "authorization:Bearer $(gcloud auth print-access-token)"  -X GET https://www.googleapis.com/oauth2/v1/userinfo\?alt\=json | jq .id`.
2. Make sure the deployed auth proxy has you on the whitelist, you could check by: `curl -X GET "https://cromwell-gotc-auth.gotc-dev.broadinstitute.org/api/workflows/v1/query?status=Running" -H "accept: application/json" -H "authorization:Bearer $(gcloud auth print-access-token)"`, if not authorized, follow the steps while on the VPN/Broad network:
    - `helm repo add gotc-charts https://broadinstitute.github.io/gotc-helm-repo/`
    - `helm list` (sanity check the list)
    - `helm upgrade gotc-dev gotc-charts/authproxy -f custom.yaml`, the custom yaml should be rendered from https://github.com/broadinstitute/gotc-deploy/blob/master/deploy/gotc-dev/helm/cromwell-auth_proxy-values.yaml.ctmpl, with the right `user-id`s in it.
3. with `./ops/server.sh` up and running, run something like `curl -X GET "http://localhost:3000/api/v1/statuscounts?environment=gotc-dev" -H "accept: application/json" -H "authorization:Bearer $(gcloud auth print-access-token)"`

---

Alternatively, after **step 1.**, you could run:
```
curl -X GET "https://workflow-launcher.gotc-dev.broadinstitute.org/api/v1/statuscounts?environment=wgs-dev" -H "accept: application/json" -H "authorization:Bearer $(gcloud auth print-access-token)"
```
and you could see the results returned all the way back from Cromwell to WFL API.
